lib/awful/screenshot.lua

```diff
@@ -0,0 +1,870 @@
+  img:finish()
+
+  frame.visible = true
+
+end
+
+-- Internal function that generates the callback to be passed to the
+-- mousegrabber that implements the snipper.
+--
+-- The snipper tool is basically a mousegrabber, which takes a single function
+-- of one argument, representing the mouse state data. This is a simple
+-- starting point that hard codes the snipper tool as being a two press design
+-- using button 1, with button 3 functioning as the cancel. These aspects of
+-- the interface can be made into parameters at some point passed as arguments.
+local function mk_mg_callback(ss)
+
+  ss._private.mg_first_pnt = {}
+
+  local function ret_mg_callback(mouse_data)
+
+    if mouse_data["buttons"][3] then
+      if ss._private.frame and ss._private.frame.visible then
+        show_frame(ss)
+      end
+      ss._private.mg_first_pnt = nil
+      ss._private.frame = nil
+      return false
+    end
+  
+    if ss._private.mg_first_pnt[1] then
+  
+      local min_x, max_x, min_y, max_y
+      min_x = math.min(ss._private.mg_first_pnt[1], mouse_data["x"])
+      max_x = math.max(ss._private.mg_first_pnt[1], mouse_data["x"])
+      min_y = math.min(ss._private.mg_first_pnt[2], mouse_data["y"])
+      max_y = math.max(ss._private.mg_first_pnt[2], mouse_data["y"])
+  
+      -- Force a minimum size to the box
+      if  max_x - min_x < 4 or max_y - min_y < 4 then
+        if frame and frame.visible then
+          show_frame(ss)
+        end
+      elseif not mouse_data["buttons"][1] then
+        show_frame(ss, {x = min_x, y = min_y, width = max_x - min_x, height = max_y - min_y})
+      end
+  
+      if mouse_data["buttons"][1] then
+  
+        local snip_surf
+        local date_time
+  
+        if ss._private.frame and ss._private.frame.visible then
+          show_frame(ss)
+        end
+  
+        ss._private.frame = nil
+        ss._private.mg_first_pnt = nil
+  
+        -- This may fail gracefully anyway but require a minimum 3x3 of pixels
+        if min_x >= max_x-1 or min_y >= max_y-1 then
+          return false
+        end
+  
+        snip_surf = crop_shot(min_x, min_y, max_x - min_x, max_y - min_y)
+        ss:filepath_builder()
+        ss._private.surface = gears.surface(snip_surf) -- surface has no setter
+  
+        if ss._private.on_success_cb then
+          ss._private.on_success_cb(ss)
+        end
+  
+        return false
+  
+      end
+  
+    else
+      if mouse_data["buttons"][1] then
+        ss._private.mg_first_pnt[1] = mouse_data["x"]
+        ss._private.mg_first_pnt[2] = mouse_data["y"]
+      end
+    end
+  
+    return true
+
+  end
+
+  return ret_mg_callback
+
+end
+
+-- Internal function to be passed as the default callback upon completion of
+-- the mousgrabber for the snipper if the user does not pass one.
+function default_on_success_cb(ss)
+  ss:save()
+end
+
+-- Internal function exected when a root window screenshot is taken.
+function screenshot_methods.root(ss)
```

Comment:
please replace `screenshot_methods.root(ss)` with `screenshot_methods:root()`, then use `self` instead of `ss`. This is mostly because it's done like this elsewhere in the codebase. Please also change it in other places too.

Reply:
These are internal functions to the library that actually take the screenshot. We have been going around in circles for almost a year now, and I am happy to design the library to whatever organization you want, but in the absence of that I had to make my own decisions. The high level way it currently works is:

``
local ss = awful.screenshot(args)
ss:save()
``

The first line calls `module.new(_, args)` as you described to me. The various parameters of the screenshot are controlled  with `args`. There are helper routines that remove the need to pass the method as an argument:

```
local ss = awful.screenshot.root(args)
ss:save()
```

These helper routines call `module()` which means it calls `new`. If no additional arguments need to be passed beyond the method:

```
local ss = awful.screenshot.root()
ss:save()
```

Works fine (this is what I do in my `rc.lua`). Please let me know if you want something different.

Reply:
The functions you're talking about here are `module.root(args)` (and similar) from further down in the file. Elv is commenting on `screenshot_methods.root(ss)`, which would map to `ss:root()` in your snippets.

`ss` here looks very much like an instance of something (methods called on it, `_private` accessed). By convention, we use Lua's `:` shorthand for functions that take an object instance as first parameter `self`.